phep-proposal.md

```diff
@@ -0,0 +1,119 @@
+
+Note that old Pharo parser will not be able to parse new numeric literal with `_`.
+No specific plan is proposed to address this issue except updating the image with a newer NumberParser.
+Numeric objects and compiled methods are unaffected (since the internal object representation of numeric does not change).
+
+## Constraints on the underscore
+
+### In other languages
+
+Languages like Java, Python or Ruby add some constraints on the placement of `_` (with subtle differences between them).
+For instance, a common rule is no trailing `_`.
+
+```
+$ echo 'class X { int x = 1_; }' > X.java && javac X.java
+X.java:1: error: illegal underscore
+class X { int x = 1_; }
+                   ^
+1 error
+
+$ python3 -c 'print(1_)'
+  File "<string>", line 1
+    print(1_)
+           ^
+SyntaxError: invalid decimal literal
+
+$ ruby -e 'print(1_)'
+-e:1: trailing `_' in number
+print(1_)
+```
+
+In these languages, such errors are HARD errors and correspond to a invalid numeric literal (not just a random illegal character).
+
+### Proposal for Pharo
+
+Instead of a formal language definition or a list on complex rules, the following single rule is considered: LEGAL=*`_` is legal only between 2 digits*.
+
+This means that leading, trailing and duplicated `_` are forbidden including around special characters `r`, `.` and `e`.
+
+* Examples of valid literals: `1`. `1_1`. `1_1.1_1`. `1_1r1_1.1_1e1_1`.
+* Examples of invalid literals: `_1`. `1_`. `1__1`. `1_.1`. `1_e1`.
+
+Here a discussion on  behaviors for Pharo are:
+
+1. Implements LEGAL in the NumberParser class but still follows the principle of the longest valid numeric token.
+
+   `1_1_a` will be parsed as `1_1` (longest valid literal according to LEGAL) followed by `_a` ie `11` (integer) and `_a` (unary message). No error or warning message.
+
+   This behavior is error prone since weird syntax will be unexpectedly and silently parsed as "something".
+   However, this is the closest thing to the current behavior according to the current implementation. (e.g. can you guess how `16rDeADMAN1` is currently parsed?)
+   Syntax highlighting (both number literal and unknown message) and fast runtime error is expected to help the programmer to deal with these cases.
+
+2. Implements LEGAL in the NumberParser class but introduce an alternative illegal numeric literal as possible token.
+
+  `1_1_a` will be parser as an error token `1_1_` ("numeric with trailing `_`") followed by the message send `a`.
+
+   This behavior is the closest to the other considered languages (Java, Python, Ruby).
+
+   Currently, `NumberParser`usually try to parses until it can't. In most case, the caller either accept the longest valid numeric literal, or complains if the whole input is not used.
```

Comment:
```suggestion
   Currently, `NumberParser`usually tries to parse until it can't. In most cases, the caller either accepts the longest valid numeric literal or complains if the whole input is not used.
```